zai_sapi/testing/catch2.hpp

```diff
@@ -48,39 +65,169 @@ extern "C" {
     __ZAI_SAPI_TEST_CASE_SUITE " [" __ZAI_SAPI_TEST_CASE_DESCRIPTION "]", \
     "[" __ZAI_SAPI_TEST_CASE_SUITE "]" __ZAI_SAPI_TEST_CASE_TAGS /* }}} */
 
+/* {{{ Test Case Fixing */
+typedef enum {
+    ZAI_SAPI_TEST_CASE_STAGE_INITIAL =  0b000,
+    ZAI_SAPI_TEST_CASE_STAGE_PROLOGUE = 0b001,
+    ZAI_SAPI_TEST_CASE_STAGE_PREFORK =  0b010,
+    ZAI_SAPI_TEST_CASE_STAGE_REQUEST =  0b100
+} ZaiSapiTestCaseStage;
+
+class ZaiSapiTestCaseFixture {
+public:
+    ZaiSapiTestCaseFixture() {
+       stage = ZAI_SAPI_TEST_CASE_STAGE_INITIAL;
+    }
+
+    bool zai_sapi_sinit() {
+        if (::zai_sapi_sinit()) {
+            stage |= ZAI_SAPI_TEST_CASE_STAGE_PROLOGUE;
+            return true;
+        }
+        return false;
+    }
+
+    bool zai_sapi_minit() {
+        if (::zai_sapi_minit()) {
+            stage |= ZAI_SAPI_TEST_CASE_STAGE_PREFORK;
+            return true;
+        }
+        return false;
+    }
+
+    bool zai_sapi_rinit() {
+        if (::zai_sapi_rinit()) {
+            stage |= ZAI_SAPI_TEST_CASE_STAGE_REQUEST;
+            return true;
+        }
+        return false;
+    }
+
+    bool zai_sapi_spinup() {
+        return zai_sapi_sinit() &&
+               zai_sapi_minit() &&
+               zai_sapi_rinit();
+    }
+
+    void zai_sapi_rshutdown() {
+        ::zai_sapi_rshutdown();
+
+        stage &= ~ZAI_SAPI_TEST_CASE_STAGE_REQUEST;
```

Comment:
I think we should swap these lines, a double shutdown is probably much worse than a partial one (if there are shutdown some requires executed during shutdown or similar)

Reply:
ack